File: apps/backend/src/executors/restExecutor.ts
export interface RestExecutorInput {
  baseUrl:        string;
  authType:       string | null;
  resolvedSecret: string | null;
  endpoint:       string;
  method:         string;
  params?:        Record<string, unknown>;
  body?:          Record<string, unknown>;
}

export interface ExecutorResult {
  success: boolean;
  data?:   unknown;
  error?:  string;
}

const TIMEOUT_MS = 30_000;

export async function restExecutor(input: RestExecutorInput): Promise<ExecutorResult> {
  const { baseUrl, authType, resolvedSecret, endpoint, method, params, body } = input;

  // Build URL — strip trailing slash from base, ensure leading slash on endpoint
  const base = baseUrl.replace(/\/$/, '');
  const path = endpoint.startsWith('/') ? endpoint : `/${endpoint}`;
  const url  = new URL(`${base}${path}`);

  if (params) {
    for (const [k, v] of Object.entries(params)) {
      url.searchParams.set(k, String(v));
    }
  }

  // Apply authentication
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'Accept':       'application/json',
  };

  if (resolvedSecret) {
    switch (authType) {
      case 'bearer':
        headers['Authorization'] = `Bearer ${resolvedSecret}`;
        break;
      case 'api_key':
        headers['X-API-Key'] = resolvedSecret;
        break;
      case 'basic':
        headers['Authorization'] = `Basic ${Buffer.from(resolvedSecret).toString('base64')}`;
        break;
      // 'none' and unknown: no auth header
    }
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);

  try {
    const hasBody = body !== undefined && method !== 'GET' && method !== 'HEAD';

    const response = await fetch(url.toString(), {
      method,
      headers,
      body:   hasBody ? JSON.stringify(body) : undefined,
      signal: controller.signal,
    });

    // Parse response body — accept JSON or fall back to text
    const ct   = response.headers.get('content-type') ?? '';
    const data = ct.includes('application/json')
      ? await response.json()
      : await response.text();

    if (!response.ok) {
      return {
        success: false,
        error:   `Upstream returned ${response.status} ${response.statusText}`,
        data,
      };
    }

    return { success: true, data };
  } catch (err) {
    if ((err as Error).name === 'AbortError') {
      return { success: false, error: 'Request timed out after 30 seconds' };
    }
    return { success: false, error: (err as Error).message };
  } finally {
    clearTimeout(timer);
  }
}

File: apps/backend-py/app/executors/rest_executor.py
import base64
from typing import Any

import httpx

TIMEOUT_S = 30.0


def _build_headers(auth_type: str | None, resolved_secret: str | None) -> dict[str, str]:
    headers = {"Content-Type": "application/json", "Accept": "application/json"}
    if not resolved_secret:
        return headers
    if auth_type == "bearer":
        headers["Authorization"] = f"Bearer {resolved_secret}"
    elif auth_type == "api_key":
        headers["X-API-Key"] = resolved_secret
    elif auth_type == "basic":
        encoded = base64.b64encode(resolved_secret.encode()).decode()
        headers["Authorization"] = f"Basic {encoded}"
    return headers


async def rest_executor(
    base_url: str,
    auth_type: str | None,
    resolved_secret: str | None,
    endpoint: str,
    method: str,
    params: dict[str, Any] | None = None,
    body: dict[str, Any] | None = None,
) -> dict[str, Any]:
    base = base_url.rstrip("/")
    path = endpoint if endpoint.startswith("/") else f"/{endpoint}"
    url = f"{base}{path}"

    headers = _build_headers(auth_type, resolved_secret)
    has_body = body is not None and method not in ("GET", "HEAD")

    try:
        async with httpx.AsyncClient(timeout=TIMEOUT_S) as client:
            resp = await client.request(
                method=method,
                url=url,
                headers=headers,
                params={k: str(v) for k, v in (params or {}).items()},
                json=body if has_body else None,
            )

        ct = resp.headers.get("content-type", "")
        try:
            data: Any = resp.json() if "application/json" in ct else resp.text
        except Exception:
            data = resp.text

        if not resp.is_success:
            detail = ""
            if isinstance(data, dict):
                detail = str(data.get("detail") or data.get("error") or data.get("message") or "")[:200]
            elif isinstance(data, str):
                detail = data[:200]
            error_msg = (
                f"Upstream returned {resp.status_code} {resp.reason_phrase} — {detail}"
                if detail
                else f"Upstream returned {resp.status_code} {resp.reason_phrase}"
            )
            return {"success": False, "error": error_msg, "data": data}

        return {"success": True, "data": data}

    except httpx.TimeoutException:
        return {"success": False, "error": "Request timed out after 30 seconds"}
    except Exception as exc:
        return {"success": False, "error": str(exc)}
